core/src/main/scala/org/apache/spark/sql/execution/columnar/encoding/DictionaryEncoding.scala

```diff
@@ -106,96 +118,46 @@ abstract class DictionaryDecoderBase
       case _ => throw new UnsupportedOperationException(
         s"DictionaryEncoding not supported for $dataType")
     }
-    baseCursor = position
     position
   }
 
-  private[sql] def initializeBeforeFinish(dictionaryBytes: AnyRef,
-      dictionaryCursor: Long, dictionaryLen: Int, cursor: Long, dataType: DataType): Unit = {
-    initializeDictionary(dictionaryBytes, dictionaryCursor,
-      dictionaryLen, dictionaryLen, dataType)
-    // base cursor will start at the given cursor since dictionary is separate
-    baseCursor = cursor
-  }
-
-  override def nextUTF8String(columnBytes: AnyRef, cursor: Long): Long =
-    cursor + 2
-
-  override def absoluteUTF8String(columnBytes: AnyRef, position: Int): Long = {
-    // TODO: PERF: optimize for local index access case by filling
-    // in the dictionary array lazily on access
-    baseCursor + ((position - numNullsUntilPosition(columnBytes, position)) << 1)
-  }
-
-  override def readUTF8String(columnBytes: AnyRef, cursor: Long): UTF8String =
-    stringDictionary.getString(columnBytes, ColumnEncoding.readShort(columnBytes, cursor))
+  // TODO: PERF: optimize for local index access case by filling
+  // in the dictionary array lazily on access
+  override def readUTF8String(columnBytes: AnyRef, nonNullPosition: Int): UTF8String =
+  stringDictionary.getString(columnBytes,
```

Comment:
> what is the overhead of creation of objects in every call?

If the column is mostly unique then object will be created at most once or not at all if there are pushed down filters. Multiple creation cases will be the in-between ones say 10K unique in total of 100K strings.

> Say for e.g. after every 1000 entries, rollover to a new dictionary.

How will this help in reducing the lifetime of objects? We will have to keep all the dictionaries alive for the whole duration of batch processing, no? Perhaps some reduction in first creation if we populate dictionaries lazily as I have mentioned in a TODO in the code. Is that what you have in mind and will be enough for the case mentioned above (10K dictionary items for total of 100K)?